File: leetcode/algorithms/344_reverse_string/main.js
/**
 * Two pointers
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(1)
 */
/**
 * @param {character[]} s
 * @return {void} Do not return anything, modify s in-place instead.
 */
var reverseString = function (s) {
  for (let i = 0; i < s.length / 2; i++) {
    const temp = s[i];
    s[i] = s[s.length - 1 - i];
    s[s.length - 1 - i] = temp;
  }
};


// Solution
/**
 * Solution 1
 *
 * Two pointers
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(1)
 */
/**
 * @param {character[]} s
 * @return {void} Do not return anything, modify s in-place instead.
 */
var solution1 = function (s) {
  let left = 0;
  let right = s.length - 1;

  while (left < right) {
    [s[left], s[right]] = [s[right], s[left]];
    left++;
    right--;
  }
};

/**
 * Solution 2
 *
 * Two pointers
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(1)
 */
/**
 * @param {character[]} s
 * @return {void} Do not return anything, modify s in-place instead.
 */
var solution1 = function (s) {
  for (let left = 0, right = s.length - 1; left < right; left++, right--) {
    const temp = s[left];
    s[left] = s[right];
    s[right] = temp;
  }
};

/**
 * Solution 3
 *
 * Built-in function
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(1)
 */
/**
 * @param {character[]} s
 * @return {void} Do not return anything, modify s in-place instead.
 */
var solution3 = function (s) {
  s.reverse();
};

/**
 * Solution 4
 *
 * Recursion
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(N)
 */
/**
 * @param {character[]} s
 * @return {void} Do not return anything, modify s in-place instead.
 */
var solution4 = function (s) {
  if (left >= right) {
    return;
  }

  [s[left], s[right]] = [s[right], s[left]];

  solution4(s, left + 1, right - 1);
};

File: leetcode/algorithms/344_reverse_string/main.java
class ReverseString {
    /**
     * Two pointers
     *
     * Complexities:
     *   N - The size of `s`
     *   - Time Complexities: O(N)
     *   - Space Complexities: O(1)
     */
    public void reverseString(char[] s) {
        for (int i = 0; i < s.length / 2; i++) {
            char temp = s[i];
            s[i] = s[s.length - 1 - i];
            s[s.length - 1 - i] = temp;
        }
    }


    // Solution
    /**
     * Solution 1
     *
     * Two pointers
     *
     * Complexities:
     *   N - The size of `s`
     *   - Time Complexities: O(N)
     *   - Space Complexities: O(1)
     */
    public void solution1(char[] s) {
        int left = 0, right = s.length - 1;

        while (left < right) {
            char tmp = s[left];
            s[left] = s[right];
            s[right] = tmp;
            left++;
            right--;
        }
    }

    /**
     * Solution 2
     *
     * Two pointers
     *
     * Complexities:
     *   N - The size of `s`
     *   - Time Complexities: O(N)
     *   - Space Complexities: O(1)
     */
    public void solution2(char[] s) {
        for (int i = 0, j = s.length - 1; i < j; i++, j--) {
            char tmp = s[i];
            s[i] = s[j];
            s[j] = tmp;
        }
    }

    /**
     * Solution 3
     *
     * Recursion
     *
     * Complexities:
     *   N - The size of `s`
     *   - Time Complexities: O(N)
     *   - Space Complexities: O(N)
     */
    public void solution3(char[] s) {
        helper(s, 0, s.length - 1);
    }

    private void helper(char[] s, int left, int right) {
        if (left >= right) {
            return;
        }

        char tmp = s[left];
        s[left] = s[right];
        s[right] = tmp;
        helper(s, left + 1, right - 1);
    }
}
